### PR TITLE
fix(nitro): skip copying symlinks and directories

### PR DESCRIPTION
**Pull request by antfu**

### CODE REVIEW

File: packages/nitro/src/rollup/plugins/externals.ts [hunk 1]
```diff
@@ -86,6 +86,9 @@ export function externals (opts: NodeExternalsOptions): Plugin {
         }
 
         const writeFile = async (file) => {
+          if (await fse.stat(file).then(i => i.isDirectory())) {
```
[the remark this answers is not in the document]
Comment (antfu) [reply packages/nitro/src/rollup/plugins/externals.ts:89]: Here I created another repro using only the start template: https://github.com/antfu/nuxt3-pnpm-nft-repro

When you build, the error:

```bash
 ERROR  Rollup error: EISDIR: illegal operation on a directory, copyfile '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ohmyfetch@0.4.2/node_modules/ufo' -> '/Users/antfu/i/_temp/nuxt3-app/.output/server/node_modules/ufo'


 ERROR  EISDIR: illegal operation on a directory, copyfile '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ohmyfetch@0.4.2/node_modules/ufo' -> '/Users/antfu/i/_temp/nuxt3-app/.output/server/node_modules/ufo'
```

And if I directly edit the nitro code in node_modules to log out `tracedFiles`, here is what we got:

<details>

<summary>Log</summary>

```js
[                                                                                                      18:56:43
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/mock/proxy.cjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/fetch/index.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ufo@0.7.9/node_modules/ufo/dist/index.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/destr@1.1.0/node_modules/destr/dist/index.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/pathe@0.2.0/node_modules/pathe/dist/index.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/h3@0.3.3/node_modules/h3/dist/index.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/polyfill/fetch.node.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/mock/proxy.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/defu@5.0.0/node_modules/defu/dist/defu.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ohmyfetch@0.4.2/node_modules/ohmyfetch/dist/node.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/destr@1.1.0/node_modules/destr/package.json',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/polyfill/globalThis.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/fetch/call.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ohmyfetch@0.4.2/node_modules/ohmyfetch/dist/chunks/fetch.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/node/http/request.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/node/http/response.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ohmyfetch@0.4.2/node_modules/destr',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ohmyfetch@0.4.2/node_modules/node-fetch',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ohmyfetch@0.4.2/node_modules/ufo',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/ufo@0.7.9/node_modules/ufo/package.json',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/node/net/socket.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/node/stream/readable.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/_internal/utils.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/node/stream/writable.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/node-fetch',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/unenv@0.4.0/node_modules/unenv/runtime/node/stream/duplex.mjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/package.json',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/index.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/body.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/response.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/request.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/headers.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/errors/abort-error.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/utils/is-redirect.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/errors/fetch-error.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/utils/is.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/utils/get-search.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/errors/base.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/node-fetch/src/utils/form-data.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/data-uri-to-buffer',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/fetch-blob',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/data-uri-to-buffer@3.0.1/node_modules/data-uri-to-buffer/package.json',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/fetch-blob@3.1.3/node_modules/fetch-blob/package.json',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/fetch-blob@3.1.3/node_modules/fetch-blob/index.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/data-uri-to-buffer@3.0.1/node_modules/data-uri-to-buffer/dist/src/index.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/fetch-blob@3.1.3/node_modules/fetch-blob/streams.cjs',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/fetch-blob@3.1.3/node_modules/web-streams-polyfill',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/web-streams-polyfill@3.1.1/node_modules/web-streams-polyfill/dist/ponyfill.es2018.js',
  '/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/web-streams-polyfill@3.1.1/node_modules/web-streams-polyfill/package.json'
]
```

</detials>

Some fo the package dirs is sneaked into the list, for example, `/Users/antfu/i/_temp/nuxt3-app/node_modules/.pnpm/node-fetch@3.0.0/node_modules/data-uri-to-buffer`.

[the remark this answers is not in the document]
Comment (pi0) [reply packages/nitro/src/rollup/plugins/externals.ts:89]: Thanks! So i guess we can improve logic instead to support `.pnpm/[id]` as package identifier.

[the remark this answers is not in the document]
Comment (antfu) [reply packages/nitro/src/rollup/plugins/externals.ts:89]: I have tried a few solutions, but actually, this implementation works the best. The reason why the dir is included in the traced files is that they are actually symbol links to the root packages.

image

The current approach to split `node_modules/` and pop already grab the last package name and makes the output node_modules works like hoist.

[the remark this answers is not in the document]
Comment (pi0) [reply packages/nitro/src/rollup/plugins/externals.ts:89]: Ok makes sense. I've added a comment to ask if it is intended or not with nft (https://github.com/vercel/nft/pull/238#discussion_r741003662)

**Pull request merged**